### PR TITLE
User management API

### PR DESCRIPTION
**Pull request by prashbnair**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,9 +1,9 @@
 FROM registry.centos.org/centos/centos:7
 
 ENV LANG=en_US.UTF-8 \
-    F8A_WORKER_VERSION=d403113 \
+    F8A_WORKER_VERSION=58d3025 \
     F8A_AUTH_VERSION=5211e23 \
-    F8A_UTILS=f94a04e
+    F8A_UTILS=3bca34e
```
Comment (yzainee) [inline Dockerfile:6]: Any specific reasons why utils is updated?

File: bayesian/api/user_api.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+from fabric8a_auth.errors import AuthError
+
+from bayesian.exceptions import HTTPError
+from bayesian.utility import user_utils
+from bayesian.utility.user_utils import UserException
+
+user_api = Blueprint('user_api', __name__, url_prefix='/user')
+
+
+@user_api.route('/readiness')
```
Comment (yzainee) [inline bayesian/api/user_api.py:17]: why do we need readiness and liveness probe here? Its for the pod health check.

File: bayesian/api/user_api.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+@user_api.route('/<user_id>', methods=['GET'])
+@login_required
+def get_user(user_id):
+    """Endpoint for getting user details."""
+    if not user_id:
+        raise HTTPError(400, "user id should be present")
+
+    user = user_utils.get_user(user_id)
+    return jsonify(user_id=user.user_id, snyk_api_token=user.snyk_api_token)
```
Comment (yzainee) [inline bayesian/api/user_api.py:37]: what happens when snyk_api_key is not present? wont user.snyk_api_key throw exception?

File: bayesian/utility/user_utils.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def get_user(uuid_val):
+    """Get User."""
+    try:
+        query = rdb.session.query(UserDetails).filter(UserDetails.user_id == uuid_val)
+        user = query.one()
+        return user
+    except SQLAlchemyError:
+        rdb.session.rollback()
```
Comment (yzainee) [inline bayesian/utility/user_utils.py:21]: I dont think we need a rollback here

File: bayesian/utility/user_utils.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        logger.exception(f"Error fetching user with id {uuid_val}")
+        raise UserException("Error fetching user")
+
+
+def create_user(uuid_val, user_source):
+    """Create User."""
+    try:
+        insert_user_stmt = insert(UserDetails).values(user_id=uuid_val, user_source=user_source,
```
Comment (yzainee) [inline bayesian/utility/user_utils.py:29]: I think you should add strftime to add the date in proper format that we want.

File: bayesian/api/user_api.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@user_api.route('', methods=['POST'])
+@login_required
+def create_user():
+    """Endpoint for creating user details."""
+    user_uuid = uuid.uuid4()
+    user_utils.create_user(user_uuid, "SNYK")
```
Comment (yzainee) [inline bayesian/api/user_api.py:45]: In case of exception while inserting user in rds, we will still be returning the uuid. Is it intended?

File: bayesian/utility/user_utils.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        rdb.session.rollback()
+        logger.exception("Error creating new user")
+        raise UserException("Error creating user")
+
+
+def update_user(user_id, snyk_api_token):
+    """Update User."""
+    try:
+        update_stmt = update(UserDetails).where(UserDetails.user_id == user_id).values(
+            dict(snyk_api_token=snyk_api_token, updated_date=datetime.datetime.now(),
```
Comment (yzainee) [inline bayesian/utility/user_utils.py:45]: same comment related to date value as mentioned above

File: bayesian/utility/user_utils.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                 registered_date=datetime.datetime.now()))
+
+        rdb.session.execute(update_stmt)
+        rdb.session.commit()
+        logger.info(f"User updated with id {user_id}")
+    except SQLAlchemyError:
+        rdb.session.rollback()
```
Comment (yzainee) [inline bayesian/utility/user_utils.py:52]: should we add some retry mechanisms here?

File: Dockerfile (hunk 1 above)
Comment (prashbnair) [reply Dockerfile:6]: I had to pull in the latest version of worker to get the RDS model. When I pulled that it required some functions that were present in a later version of utils.

File: bayesian/api/user_api.py (hunk 2 above)
Comment (prashbnair) [reply bayesian/api/user_api.py:17]: Will remove it.

File: bayesian/api/user_api.py (hunk 3 above)
Comment (prashbnair) [reply bayesian/api/user_api.py:37]: No should return None and the field will be empty.

File: bayesian/api/user_api.py (hunk 6 above)
Comment (prashbnair) [reply bayesian/api/user_api.py:45]: No, if there is an exception it is not caught and propagated up after which the exception handler will handle it. So UUID will not be returned.

File: bayesian/utility/user_utils.py (hunk 5 above)
Comment (prashbnair) [reply bayesian/utility/user_utils.py:29]: The dates seem to be in the right format, i checked using pgweb. Maybe it is doing some conversion, i will check this.

File: Dockerfile (hunk 1 above)
Comment (yzainee) [reply Dockerfile:6]: ok

File: bayesian/api/user_api.py (hunk 3 above)
Comment (yzainee) [reply bayesian/api/user_api.py:37]: ok. just test it out to make sure we dont end up in an exception

File: bayesian/utility/user_utils.py (hunk 5 above)
Comment (yzainee) [reply bayesian/utility/user_utils.py:29]: with strftime, you will be able to store it in a particular format. Then later when we want to query using these values, it will be very easy for us.

File: bayesian/api/user_api.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@user_api.route('/liveness')
+def liveness():
+    """Handle the /liveness REST API call."""
+    return jsonify({}), 200
+
+
+@user_api.route('/<user_id>', methods=['GET'])
+@login_required
```
Comment (arajkumar) [inline bayesian/api/user_api.py:30]: Can you create a API spec for these endpoints?

File: bayesian/api/user_api.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+def get_user(user_id):
```
Comment (arajkumar) [inline bayesian/api/user_api.py:31]: Can you create a API spec for these endpoints?

File: bayesian/utility/user_utils.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+def get_user(uuid_val):
+    """Get User."""
+    try:
+        query = rdb.session.query(UserDetails).filter(UserDetails.user_id == uuid_val)
+        user = query.one()
+        return user
+    except SQLAlchemyError:
+        rdb.session.rollback()
+        logger.exception(f"Error fetching user with id {uuid_val}")
+        raise UserException("Error fetching user")
```
Comment (arajkumar) [inline bayesian/utility/user_utils.py:23]: Follow `raise Exception(...) from e` pattern to preserve actual call trace when rethrowing an exception

File: bayesian/utility/user_utils.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+from f8a_worker.models import (UserDetails)
+from bayesian import rdb
+
+logger = logging.getLogger(__file__)
+
+
+def get_user(uuid_val):
+    """Get User."""
+    try:
+        query = rdb.session.query(UserDetails).filter(UserDetails.user_id == uuid_val)
```
Comment (arajkumar) [inline bayesian/utility/user_utils.py:17]: I guess `query` has to be renamed to `result`.?

File: tests/utility/test_user_utils.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
+"""Test Case for User Management Utilities."""
```
Comment (arajkumar) [inline tests/utility/test_user_utils.py:1]: This seems to be redundant as you already test user_utils through user_api.

File: bayesian/utility/user_utils.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+class UserException(Exception):
+    """Exception for all User Management."""
+
+    def __init__(self, message):
+        """Initialize the exception."""
+        self.message = message
+
+    def __str__(self):
+        """Representation of the exception."""
+        return self.message
```
Comment (arajkumar) [inline bayesian/utility/user_utils.py:66]: This is not required, you can access arguments to custom exception through exception_obj.args field.

File: bayesian/api/user_api.py (hunk 10 above)
Comment (prashbnair) [reply bayesian/api/user_api.py:31]: Sure, will do

File: bayesian/utility/user_utils.py (hunk 12 above)
Comment (prashbnair) [reply bayesian/utility/user_utils.py:17]: Done

File: bayesian/utility/user_utils.py (hunk 4 above)
Comment (prashbnair) [reply bayesian/utility/user_utils.py:21]: Done

File: tests/utility/test_user_utils.py (hunk 13 above)
Comment (prashbnair) [reply tests/utility/test_user_utils.py:1]: No, I am mocking out the utility in the api tests.

File: bayesian/utility/user_utils.py (hunk 11 above)
Comment (prashbnair) [reply bayesian/utility/user_utils.py:23]: Done

File: bayesian/utility/user_utils.py (hunk 5 above)
Comment (prashbnair) [reply bayesian/utility/user_utils.py:29]: This converts it to a string but the data type in the DB is datetime.  If we convert to string we will loose the benefits that the DB provides with datetime.

Comment (yzainee) [reply bayesian/utility/user_utils.py:29]: oh ok. schema is datetime. then we cannot use strftime.

File: bayesian/api/user_api.py (hunk 10 above)
Comment (prashbnair) [reply bayesian/api/user_api.py:31]: https://app.swaggerhub.com/apis/prashbnair/User/1.0.0#/

File: tests/utility/test_user_utils.py (hunk 13 above)
Comment (arajkumar) [reply tests/utility/test_user_utils.py:1]: Don't mock utility from api test and mock only external sources(DB).

File: bayesian/api/user_api.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    """Endpoint for getting user details."""
+    if not user_id:
+        raise HTTPError(400, "user id should be present")
+
+    user = user_utils.get_user(user_id)
+    return jsonify(user_id=user.user_id, snyk_api_token=user.snyk_api_token)
```
Comment (arajkumar) [inline bayesian/api/user_api.py:25]: What is the purpose of returning `snyk_api_token` through API? Is it just to check whether the UUID is associated with registered snyk account? If so, can abstract it with `registration_status` instead of `snyk_api_token`?

File: tests/utility/test_user_utils.py (hunk 13 above)
Comment (prashbnair) [reply tests/utility/test_user_utils.py:1]: Isnt unit testing about testing the unit? Thats why we have fixtures and mocks If there was no external source then we would only need to test the API and dont need to write test cases for anything else. That is not correct.

File: bayesian/api/user_api.py (hunk 15 above)
Comment (prashbnair) [reply bayesian/api/user_api.py:25]: Yes this can be changed to status

File: tests/utility/test_user_utils.py (hunk 13 above)
Comment (arajkumar) [reply tests/utility/test_user_utils.py:1]: Unit testing doesn't mean we have to have test files for individual classes. We need to test the module/component as a whole by mocking out the external boundaries. If we have unittest per class/per function, it would definitely hinder ability to refactor the code. Excessive mocking cause refactoring hard. There is a very good post on this. Please check: https://medium.com/@Cyrdup/unit-testing-youre-doing-it-wrong-407a07692989

Comment (prashbnair) [reply tests/utility/test_user_utils.py:1]: There are several opinions on unit testing. TDD states that every function needs to first have a test case before it is written. It also says that you should test your function without testing the dependencies. This will make sure that your function is doing what is intended.

File: bayesian/utility/user_utils.py (hunk 8 above)
Comment (yzainee) [reply bayesian/utility/user_utils.py:53]: Are we not going to add any retry mech here? If the operation fails due to some timeout etc, should we not have a retry thing for say 3 times before throwing an exception?

File: bayesian/api/user_api.py (hunk 6 above)
Comment (yzainee) [reply bayesian/api/user_api.py:33]: what would be the response that will be sent back to the API caller?

File: tests/utility/test_user_utils.py (hunk 13 above)
Comment (arajkumar) [reply tests/utility/test_user_utils.py:1]: AFAIK, TDD focuses on test first approach(i.e. define tests to cover the spec) and write minimal functionality to pass the test and improve upon with further test cases. It never defines what is unit. Unit is subjective and doesn't mean always a class/function.

Comment (prashbnair) [reply tests/utility/test_user_utils.py:1]: When you do TDD, you focus on just the function that you are working on. It encourages you to not bother about other dependencies.  Unit testing like I said even earlier has not defined what a unit but there are multiple ways of doing it so my suggestion is go through the test cases that have been written. If you do not find value in them, it can be discussed. If you want to enforce something and it is being followed all across the project I am fine with that also.

File: bayesian/utility/user_utils.py (hunk 8 above)
Comment (prashbnair) [reply bayesian/utility/user_utils.py:58]: Done, with tenacity library from python
